### ISSUE TITLE
keyboard_key_press() and keyboard_key_release() not implemented in SDL

### ISSUE THREAD

**Issue by time-killer-games** [enigma-dev/enigma-dev#1480]
Title says all...

**Comment by RobertBColton** [comment]
There's only one option, and I'm not sure if it sends it to other applications.
https://wiki.libsdl.org/SDL_PushEvent

Other than that, there's nothing else that can be done. This is why XTEST is an extension, use it with SDL and you have the function.

**Comment by time-killer-games** [comment]
I consider it useful. I use it to simulate keyboard events when I use the joystick functions.

**Comment by RobertBColton** [comment]
>I use it to simulate keyboard events when I use the joystick functions.

No, that's not the correct way to achieve what you want. That's not how other game engines actually do this. What you want is a map of maps. Specifically `map<string, map<int, int> >` where string is a name you give the input. This is actually how Unity's input mapping works.
https://docs.unity3d.com/Manual/ConventionalGameInput.html

I was hoping to bring this to ENIGMA at some point and it's the proper way to deal with this. You basically create a menu that listens (with some delay to not affect performance) to all input devices. That includes the keyboard, mouse, and all joystick/gamepad slots. This is a bit like what `keyboard_string` or `keyboard_wait` does, but with all devices instead of just the keyboard. Then when the user presses an input it gets mapped to the control.

Something like this is what I am proposing:
```gml
input_map_add("player_forward", im_keyboard, keyboard_lastkey);
```
You would do something like this to check input:
```gml
if (input_map_pressed("player_forward")) {
  speed = 5;
}
```

This is all hypothetical, and truth be told we don't really need to build it into the engine. Clearly it's possible to build this yourself in ENIGMA and GM on your own, using GML and `ds_map` only, and it's honestly the correct way to do it. I think we might still add it to the engine anyway for the Close/Screen Capture game settings because I want the ability to remap them to other keys in RadialGM.

**Comment by time-killer-games** [comment]
I'll just leave this up as a mental note to add it myself. I don't care if you think it isn't "the correct" way to do it. All that matters to me is that it works and it is "a" way to do it.

**Comment by RobertBColton** [comment]
Closing as resolved by #1485.
